File: count_sort.go
package main

import "fmt"

func countsort(A []int) {
	l := len(A)
	m := A[0]
	for i := 1; i < l; i++ {
		if A[i] > m {
			m = A[i]
		}
	}
	carray := make([]int, m+1)
	for i := 0; i < l; i++ {
		carray[A[i]] += 1
	}
	fmt.Println(carray)
	i := 0
	j := 0
	for i <= m {
		if carray[i] > 0 {
			A[j] = i
			j += 1
			carray[i] -= 1
		} else {
			i += 1
		}

	}
}

func main() {
	A := []int{1, 2, 3, 4, 2, 3, 1, 3}
	fmt.Println(A)
	countsort(A)
	fmt.Println(A)
}

File: count_sort.py
# count sort is a index based sort.
# counting the elements having distint key values.
# we take another array with size is equal to max value of elements to store the count of item.


def countsort(A):
    l=len(A)
    m=max(A)
    carray=[0]*(m+1)
    for i in range(l):
        # print(A[i],carray[A[i]])
        carray[A[i]]+=1
    print(carray)
    i=0
    j=0
    while i<(m+1):
        if carray[i]>0:
            A[j]=i
            j+=1
            carray[i]-=1
        else:
            i+=1



A=[1,2,3,4,2,3,1,3]
print("original :",A)
countsort(A)
print("sorted :",A)


